File: Week 22/advanced_graphs.py
"""
Week 22: Advanced Graphs
=========================
This module covers shortest-path algorithms, minimum spanning trees, and
the Disjoint Set Union (Union-Find) data structure.

Topics covered:
    1. Dijkstra's algorithm (using heapq, with path reconstruction)
    2. Bellman-Ford (with negative cycle detection)
    3. Floyd-Warshall (all-pairs shortest paths, with path reconstruction)
    4. DSU / Union-Find (path compression + union by rank)
    5. Kruskal's MST
    6. Prim's MST
"""

from __future__ import annotations

import heapq
from collections import defaultdict
from typing import Dict, List, Optional, Set, Tuple


# Type aliases for clarity
INF = float("inf")
AdjList = Dict[int, List[Tuple[int, float]]]  # vertex -> [(neighbor, weight)]


# ---------------------------------------------------------------------------
# Helper: build weighted adjacency list
# ---------------------------------------------------------------------------
def build_adj_list(
    edges: List[Tuple[int, int, float]], directed: bool = False
) -> Tuple[AdjList, Set[int]]:
    """
    Build an adjacency list from a list of (u, v, weight) edges.
    Returns (adj, vertices).
    """
    adj: AdjList = defaultdict(list)
    vertices: Set[int] = set()
    for u, v, w in edges:
        adj[u].append((v, w))
        vertices.update([u, v])
        if not directed:
            adj[v].append((u, w))
    return adj, vertices


# ---------------------------------------------------------------------------
# 1. Dijkstra's Algorithm
# ---------------------------------------------------------------------------
def dijkstra(
    adj: AdjList, source: int, vertices: Set[int]
) -> Tuple[Dict[int, float], Dict[int, Optional[int]]]:
    """
    Dijkstra's single-source shortest path algorithm for non-negative weights.

    Uses a min-heap (priority queue) to greedily process the nearest unvisited
    vertex.

    Returns:
        dist   — dict mapping each vertex to its shortest distance from source
        parent — dict for path reconstruction (parent[v] = predecessor of v)

    Time:  O((V + E) log V) with a binary heap
    Space: O(V + E)
    """
    dist: Dict[int, float] = {v: INF for v in vertices}
    parent: Dict[int, Optional[int]] = {v: None for v in vertices}
    dist[source] = 0.0

    # Min-heap: (distance, vertex)
    heap: List[Tuple[float, int]] = [(0.0, source)]

    while heap:
        d, u = heapq.heappop(heap)
        if d > dist[u]:
            continue  # stale entry — we already found a shorter path to u

        for v, w in adj[u]:
            new_dist = dist[u] + w
            if new_dist < dist[v]:
                dist[v] = new_dist
                parent[v] = u
                heapq.heappush(heap, (new_dist, v))

    return dist, parent


def reconstruct_path(parent: Dict[int, Optional[int]], target: int) -> List[int]:
    """
    Reconstruct the shortest path from source to *target* using the parent map.

    Time: O(path_length)
    """
    path: List[int] = []
    node: Optional[int] = target
    while node is not None:
        path.append(node)
        node = parent[node]
    path.reverse()
    return path


# ---------------------------------------------------------------------------
# 2. Bellman-Ford Algorithm
# ---------------------------------------------------------------------------
def bellman_ford(
    edges: List[Tuple[int, int, float]], source: int, vertices: Set[int]
) -> Tuple[Optional[Dict[int, float]], Dict[int, Optional[int]]]:
    """
    Bellman-Ford single-source shortest path algorithm.

    Unlike Dijkstra, handles *negative* edge weights.
    Detects negative-weight cycles.

    Algorithm:
        1. Initialise dist[source] = 0, all others = INF.
        2. Relax all edges V-1 times.
        3. One more pass: if any edge can still be relaxed, a negative
           cycle exists.

    Returns:
        dist   — shortest distances, or None if a negative cycle is reachable
        parent — predecessor map for path reconstruction

    Time:  O(V * E)
    Space: O(V)
    """
    dist: Dict[int, float] = {v: INF for v in vertices}
    parent: Dict[int, Optional[int]] = {v: None for v in vertices}
    dist[source] = 0.0

    n = len(vertices)

    # Relax all edges V-1 times.
    for _ in range(n - 1):
        updated = False
        for u, v, w in edges:
            if dist[u] != INF and dist[u] + w < dist[v]:
                dist[v] = dist[u] + w
                parent[v] = u
                updated = True
        if not updated:
            break  # early termination — no changes in this pass

    # Check for negative-weight cycles.
    for u, v, w in edges:
        if dist[u] != INF and dist[u] + w < dist[v]:
            return None, parent  # negative cycle detected

    return dist, parent


# ---------------------------------------------------------------------------
# 3. Floyd-Warshall (all-pairs shortest paths)
# ---------------------------------------------------------------------------
def floyd_warshall(
    n: int, edges: List[Tuple[int, int, float]]
) -> Tuple[List[List[float]], List[List[Optional[int]]]]:
    """
    Floyd-Warshall all-pairs shortest path algorithm.

    Vertices are numbered 0..n-1.

    Algorithm:
        For each intermediate vertex k, for each pair (i, j):
            dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j])

    Also maintains a *next* matrix for path reconstruction.

    Returns:
        dist — n x n matrix of shortest distances
        nxt  — n x n matrix where nxt[i][j] = next vertex on the shortest
               path from i to j (None if no path)

    Time:  O(V^3)
    Space: O(V^2)
    """
    dist = [[INF] * n for _ in range(n)]
    nxt: List[List[Optional[int]]] = [[None] * n for _ in range(n)]

    for i in range(n):
        dist[i][i] = 0.0

    for u, v, w in edges:
        dist[u][v] = w
        nxt[u][v] = v

    # Main triple loop
    for k in range(n):
        for i in range(n):
            for j in range(n):
                if dist[i][k] + dist[k][j] < dist[i][j]:
                    dist[i][j] = dist[i][k] + dist[k][j]
                    nxt[i][j] = nxt[i][k]

    return dist, nxt


def reconstruct_path_fw(
    nxt: List[List[Optional[int]]], u: int, v: int
) -> Optional[List[int]]:
    """
    Reconstruct the shortest path from u to v using the Floyd-Warshall
    *next* matrix.

    Returns None if no path exists.
    Time: O(path_length)
    """
    if nxt[u][v] is None:
        return None
    path = [u]
    while u != v:
        u = nxt[u][v]
        if u is None:
            return None
        path.append(u)
    return path


# ---------------------------------------------------------------------------
# 4. Disjoint Set Union (Union-Find)
# ---------------------------------------------------------------------------
class UnionFind:
    """
    Disjoint Set Union (DSU) with path compression and union by rank.

    Operations:
        find(x)      — O(alpha(n)) amortized  (nearly O(1))
        union(x, y)  — O(alpha(n)) amortized
        connected(x, y) — O(alpha(n)) amortized

    alpha is the inverse Ackermann function, which grows incredibly slowly
    and is effectively constant for all practical input sizes.

    Space: O(n)
    """

    def __init__(self, n: int) -> None:
        """Create n disjoint singleton sets {0}, {1}, ..., {n-1}."""
        self.parent = list(range(n))
        self.rank = [0] * n
        self.count = n  # number of disjoint sets

    def find(self, x: int) -> int:
        """
        Find the representative (root) of the set containing *x*.
        Uses path compression: every node on the path to the root is made
        to point directly at the root, flattening the tree.
        """
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])  # path compression
        return self.parent[x]

    def union(self, x: int, y: int) -> bool:
        """
        Merge the sets containing *x* and *y*.
        Uses union by rank: attach the shorter tree under the taller one.
        Returns True if a merge happened, False if x and y were already
        in the same set.
        """
        rx, ry = self.find(x), self.find(y)
        if rx == ry:
            return False  # already in the same set

        # Union by rank
        if self.rank[rx] < self.rank[ry]:
            rx, ry = ry, rx
        self.parent[ry] = rx
        if self.rank[rx] == self.rank[ry]:
            self.rank[rx] += 1

        self.count -= 1
        return True

    def connected(self, x: int, y: int) -> bool:
        """Check if *x* and *y* are in the same set."""
        return self.find(x) == self.find(y)


# ---------------------------------------------------------------------------
# 5. Kruskal's MST
# ---------------------------------------------------------------------------
def kruskal_mst(
    n: int, edges: List[Tuple[int, int, float]]
) -> Tuple[float, List[Tuple[int, int, float]]]:
    """
    Kruskal's algorithm for Minimum Spanning Tree.

    Algorithm:
        1. Sort all edges by weight.
        2. Greedily add each edge if it doesn't form a cycle (checked via
           Union-Find).
        3. Stop after n-1 edges (MST is complete).

    Time:  O(E log E)  — dominated by sorting
    Space: O(V + E)
    """
    sorted_edges = sorted(edges, key=lambda e: e[2])
    uf = UnionFind(n)
    mst_weight = 0.0
    mst_edges: List[Tuple[int, int, float]] = []

    for u, v, w in sorted_edges:
        if uf.union(u, v):
            mst_weight += w
            mst_edges.append((u, v, w))
            if len(mst_edges) == n - 1:
                break

    return mst_weight, mst_edges


# ---------------------------------------------------------------------------
# 6. Prim's MST
# ---------------------------------------------------------------------------
def prim_mst(
    adj: AdjList, start: int, n: int
) -> Tuple[float, List[Tuple[int, int, float]]]:
    """
    Prim's algorithm for Minimum Spanning Tree using a min-heap.

    Algorithm:
        1. Start from *start* vertex.
        2. Maintain a min-heap of (weight, from_vertex, to_vertex) for
           edges crossing the cut (visited / unvisited).
        3. Greedily pick the lightest crossing edge whose endpoint is
           unvisited.

    Time:  O((V + E) log V)
    Space: O(V + E)
    """
    visited = [False] * n
    mst_weight = 0.0
    mst_edges: List[Tuple[int, int, float]] = []

    # Heap: (weight, from_vertex, to_vertex)
    heap: List[Tuple[float, int, int]] = [(0.0, -1, start)]

    while heap and len(mst_edges) < n - 1:
        w, frm, to = heapq.heappop(heap)
        if visited[to]:
            continue
        visited[to] = True
        if frm != -1:
            mst_weight += w
            mst_edges.append((frm, to, w))

        for neighbor, weight in adj[to]:
            if not visited[neighbor]:
                heapq.heappush(heap, (weight, to, neighbor))

    return mst_weight, mst_edges


# ===========================================================================
# Test Cases
# ===========================================================================
if __name__ == "__main__":
    print("=" * 60)
    print("Week 22 — Advanced Graphs")
    print("=" * 60)

    # --- Dijkstra ----------------------------------------------------------
    print("\n--- Dijkstra ---")
    edges_d = [
        (0, 1, 4), (0, 2, 1), (2, 1, 2), (1, 3, 1),
        (2, 3, 5), (3, 4, 3),
    ]
    adj_d, verts_d = build_adj_list(edges_d, directed=True)
    dist, parent = dijkstra(adj_d, 0, verts_d)
    print(f"Distances from 0: {dist}")
    assert dist[0] == 0 and dist[1] == 3 and dist[3] == 4 and dist[4] == 7

    path = reconstruct_path(parent, 4)
    print(f"Shortest path 0 -> 4: {path}")
    assert path == [0, 2, 1, 3, 4]

    # --- Bellman-Ford ------------------------------------------------------
    print("\n--- Bellman-Ford ---")
    edges_bf = [
        (0, 1, -1), (0, 2, 4), (1, 2, 3),
        (1, 3, 2), (1, 4, 2), (3, 2, 5),
        (3, 1, 1), (4, 3, -3),
    ]
    verts_bf = {0, 1, 2, 3, 4}
    dist_bf, parent_bf = bellman_ford(edges_bf, 0, verts_bf)
    assert dist_bf is not None
    print(f"Distances from 0: {dist_bf}")
    assert dist_bf[0] == 0 and dist_bf[1] == -1 and dist_bf[4] == 1

    # Negative cycle test
    edges_neg = [(0, 1, 1), (1, 2, -1), (2, 0, -1)]
    verts_neg = {0, 1, 2}
    dist_neg, _ = bellman_ford(edges_neg, 0, verts_neg)
    assert dist_neg is None
    print("Negative cycle detected correctly.")

    # --- Floyd-Warshall ----------------------------------------------------
    print("\n--- Floyd-Warshall ---")
    edges_fw = [
        (0, 1, 3), (0, 2, 8), (0, 4, -4),
        (1, 3, 1), (1, 4, 7),
        (2, 1, 4),
        (3, 0, 2), (3, 2, -5),
        (4, 3, 6),
    ]
    dist_fw, nxt_fw = floyd_warshall(5, edges_fw)
    print("Floyd-Warshall distance matrix:")
    for row in dist_fw:
        print(f"  {row}")
    assert dist_fw[0][3] == 2   # 0->4->3 = -4+6 = 2
    assert dist_fw[3][1] == -1  # 3->2->1 = -5+4 = -1

    path_fw = reconstruct_path_fw(nxt_fw, 0, 2)
    print(f"Path 0 -> 2: {path_fw}")

    # --- Union-Find --------------------------------------------------------
    print("\n--- Union-Find ---")
    uf = UnionFind(6)
    uf.union(0, 1)
    uf.union(2, 3)
    uf.union(0, 3)
    assert uf.connected(0, 2) is True
    assert uf.connected(0, 4) is False
    assert uf.count == 3  # {0,1,2,3}, {4}, {5}
    print(f"Components: {uf.count}, connected(0,2)={uf.connected(0,2)}")

    # --- Kruskal's MST -----------------------------------------------------
    print("\n--- Kruskal's MST ---")
    #   0 --1-- 1
    #   |      /|
    #   4    3  2
    #   |  /    |
    #   3 --5-- 2
    mst_edges_input = [
        (0, 1, 1), (0, 3, 4), (1, 2, 2), (1, 3, 3),
        (2, 3, 5), (2, 4, 6), (3, 4, 7),
    ]
    total, mst = kruskal_mst(5, mst_edges_input)
    print(f"MST weight: {total}, edges: {mst}")
    assert total == 12  # 1 + 2 + 3 + 6

    # --- Prim's MST --------------------------------------------------------
    print("\n--- Prim's MST ---")
    adj_p, verts_p = build_adj_list(mst_edges_input, directed=False)
    total_p, mst_p = prim_mst(adj_p, 0, 5)
    print(f"MST weight: {total_p}, edges: {mst_p}")
    assert total_p == 12

    print("\nAll tests passed!")

File: Week 22/cpp/advanced_graphs.cpp
/*
 * =============================================================================
 * Week 22 — Advanced Graphs  (C++17)
 * =============================================================================
 *
 * Topics covered
 * --------------
 *   1. Dijkstra's algorithm  (priority_queue, returns distances)
 *   2. Bellman-Ford           (with negative cycle detection)
 *   3. Floyd-Warshall         (all-pairs shortest paths)
 *   4. DSU / Union-Find class (path compression + union by rank)
 *   5. Kruskal's MST          (using DSU)
 *   6. Prim's MST             (using priority_queue)
 *
 * Complexity cheat-sheet  (V = vertices, E = edges)
 * --------------------------------------------------
 *   dijkstra          O((V + E) log V)  |  Space O(V + E)
 *   bellman_ford      O(V * E)          |  Space O(V)
 *   floyd_warshall    O(V^3)            |  Space O(V^2)
 *   DSU operations    O(alpha(n)) ~ O(1) amortized
 *   kruskal_mst       O(E log E)        |  Space O(V)
 *   prim_mst          O((V + E) log V)  |  Space O(V + E)
 *
 * Build & run
 *   g++ -std=c++17 -O2 -o advanced_graphs advanced_graphs.cpp && ./advanced_graphs
 * =============================================================================
 */

#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
#include <cassert>
#include <sstream>
#include <climits>
#include <numeric>
#include <tuple>
#include <optional>

// ---------------------------------------------------------------------------
// Helper
// ---------------------------------------------------------------------------
template <typename T>
std::string vec_str(const std::vector<T>& v) {
    std::ostringstream oss;
    oss << "[";
    for (std::size_t i = 0; i < v.size(); ++i) {
        if (i) oss << ", ";
        if constexpr (std::is_same_v<T, long long>) {
            if (v[i] == LLONG_MAX) { oss << "INF"; continue; }
        }
        oss << v[i];
    }
    oss << "]";
    return oss.str();
}

using Edge = std::tuple<int, int, int>;  // (u, v, weight)
using AdjList = std::vector<std::vector<std::pair<int, int>>>;  // adj[u] = {(v, w)}

// ---------------------------------------------------------------------------
// 1. Dijkstra — single-source shortest paths (non-negative weights)
// ---------------------------------------------------------------------------
// Complexity:  Time O((V + E) log V)  |  Space O(V + E)
std::vector<long long> dijkstra(const AdjList& adj, int src) {
    int n = static_cast<int>(adj.size());
    std::vector<long long> dist(n, LLONG_MAX);
    dist[src] = 0;

    // Min-heap: (distance, vertex)
    using P = std::pair<long long, int>;
    std::priority_queue<P, std::vector<P>, std::greater<>> pq;
    pq.push({0, src});

    while (!pq.empty()) {
        auto [d, u] = pq.top(); pq.pop();
        if (d > dist[u]) continue;   // stale entry

        for (auto [v, w] : adj[u]) {
            if (dist[u] + w < dist[v]) {
                dist[v] = dist[u] + w;
                pq.push({dist[v], v});
            }
        }
    }
    return dist;
}

// ---------------------------------------------------------------------------
// 2. Bellman-Ford — handles negative weights, detects negative cycles
// ---------------------------------------------------------------------------
// Complexity:  Time O(V * E)  |  Space O(V)
// Returns nullopt if a negative cycle is reachable from src.
std::optional<std::vector<long long>> bellman_ford(int n, const std::vector<Edge>& edges, int src) {
    std::vector<long long> dist(n, LLONG_MAX);
    dist[src] = 0;

    // Relax all edges V-1 times.
    for (int i = 0; i < n - 1; ++i) {
        for (auto [u, v, w] : edges) {
            if (dist[u] != LLONG_MAX && dist[u] + w < dist[v]) {
                dist[v] = dist[u] + w;
            }
        }
    }

    // Check for negative cycle: if we can still relax, there is one.
    for (auto [u, v, w] : edges) {
        if (dist[u] != LLONG_MAX && dist[u] + w < dist[v]) {
            return std::nullopt;  // negative cycle detected
        }
    }
    return dist;
}

// ---------------------------------------------------------------------------
// 3. Floyd-Warshall — all-pairs shortest paths
// ---------------------------------------------------------------------------
// Complexity:  Time O(V^3)  |  Space O(V^2)
constexpr long long FW_INF = 1e18;

std::vector<std::vector<long long>> floyd_warshall(int n, const std::vector<Edge>& edges) {
    std::vector<std::vector<long long>> dist(n, std::vector<long long>(n, FW_INF));

    for (int i = 0; i < n; ++i) dist[i][i] = 0;
    for (auto [u, v, w] : edges) dist[u][v] = std::min(dist[u][v], (long long)w);

    for (int k = 0; k < n; ++k)
        for (int i = 0; i < n; ++i)
            for (int j = 0; j < n; ++j)
                if (dist[i][k] < FW_INF && dist[k][j] < FW_INF)
                    dist[i][j] = std::min(dist[i][j], dist[i][k] + dist[k][j]);

    return dist;
}

// ---------------------------------------------------------------------------
// 4. DSU / Union-Find — path compression + union by rank
// ---------------------------------------------------------------------------
// Complexity:  Nearly O(1) amortized per operation (inverse Ackermann)
class DSU {
public:
    explicit DSU(int n) : parent_(n), rank_(n, 0) {
        std::iota(parent_.begin(), parent_.end(), 0);
    }

    int find(int x) {
        if (parent_[x] != x) parent_[x] = find(parent_[x]);  // path compression
        return parent_[x];
    }

    bool unite(int x, int y) {
        int px = find(x), py = find(y);
        if (px == py) return false;

        // Union by rank.
        if (rank_[px] < rank_[py]) std::swap(px, py);
        parent_[py] = px;
        if (rank_[px] == rank_[py]) ++rank_[px];
        return true;
    }

    [[nodiscard]] bool connected(int x, int y) { return find(x) == find(y); }

private:
    std::vector<int> parent_;
    std::vector<int> rank_;
};

// ---------------------------------------------------------------------------
// 5. Kruskal's MST — sort edges, use DSU
// ---------------------------------------------------------------------------
// Complexity:  Time O(E log E)  |  Space O(V)
std::pair<long long, std::vector<Edge>> kruskal_mst(int n, std::vector<Edge> edges) {
    std::sort(edges.begin(), edges.end(),
              [](const Edge& a, const Edge& b) { return std::get<2>(a) < std::get<2>(b); });

    DSU dsu(n);
    long long total = 0;
    std::vector<Edge> mst_edges;

    for (auto& [u, v, w] : edges) {
        if (dsu.unite(u, v)) {
            total += w;
            mst_edges.emplace_back(u, v, w);
            if (static_cast<int>(mst_edges.size()) == n - 1) break;
        }
    }
    return {total, mst_edges};
}

// ---------------------------------------------------------------------------
// 6. Prim's MST — priority_queue based
// ---------------------------------------------------------------------------
// Complexity:  Time O((V + E) log V)  |  Space O(V + E)
std::pair<long long, std::vector<Edge>> prim_mst(const AdjList& adj) {
    int n = static_cast<int>(adj.size());
    std::vector<bool> in_mst(n, false);
    long long total = 0;
    std::vector<Edge> mst_edges;

    // Min-heap: (weight, vertex, parent)
    using P = std::tuple<int, int, int>;
    std::priority_queue<P, std::vector<P>, std::greater<>> pq;
    pq.push({0, 0, -1});

    while (!pq.empty() && static_cast<int>(mst_edges.size()) < n) {
        auto [w, u, parent] = pq.top(); pq.pop();
        if (in_mst[u]) continue;
        in_mst[u] = true;
        total += w;
        if (parent != -1) mst_edges.emplace_back(parent, u, w);

        for (auto [v, wt] : adj[u]) {
            if (!in_mst[v]) pq.push({wt, v, u});
        }
    }
    return {total, mst_edges};
}

// ---------------------------------------------------------------------------
// main — test cases
// ---------------------------------------------------------------------------
int main() {
    std::cout << "=== Week 22: Advanced Graphs ===\n\n";

    /*
     * Test graph (directed, weighted):
     *
     *   0 --1--> 1 --3--> 3
     *   |        |        ^
     *   4        1        |
     *   v        v        1
     *   2 --5--> 3 --2--> 4  (and edge 3->4 weight 2)
     *
     * Wait, let's use a cleaner example.
     */

    // ---- 1. Dijkstra ----
    {
        std::cout << "-- Dijkstra --\n";
        int n = 5;
        AdjList adj(n);
        adj[0] = {{1, 4}, {2, 1}};
        adj[1] = {{3, 1}};
        adj[2] = {{1, 2}, {3, 5}};
        adj[3] = {{4, 3}};
        // shortest from 0: 0->2(1), 2->1(3), 1->3(4), 3->4(7)

        auto dist = dijkstra(adj, 0);
        std::cout << "  dist from 0: " << vec_str(dist) << "\n";
        assert(dist[0] == 0 && dist[1] == 3 && dist[2] == 1 && dist[3] == 4 && dist[4] == 7);
        std::cout << "\n";
    }

    // ---- 2. Bellman-Ford ----
    {
        std::cout << "-- Bellman-Ford --\n";
        int n = 5;
        std::vector<Edge> edges = {
            {0, 1, 4}, {0, 2, 1}, {2, 1, 2}, {1, 3, 1}, {2, 3, 5}, {3, 4, 3}
        };
        auto res = bellman_ford(n, edges, 0);
        assert(res.has_value());
        std::cout << "  dist from 0: " << vec_str(*res) << "\n";
        assert((*res)[4] == 7);

        // Test negative cycle detection.
        std::vector<Edge> neg_edges = {{0, 1, 1}, {1, 2, -1}, {2, 0, -1}};
        auto neg_res = bellman_ford(3, neg_edges, 0);
        assert(!neg_res.has_value());
        std::cout << "  negative cycle detected: true\n\n";
    }

    // ---- 3. Floyd-Warshall ----
    {
        std::cout << "-- Floyd-Warshall --\n";
        int n = 4;
        std::vector<Edge> edges = {
            {0, 1, 3}, {0, 3, 7}, {1, 0, 8}, {1, 2, 2},
            {2, 0, 5}, {2, 3, 1}, {3, 0, 2}
        };
        auto dist = floyd_warshall(n, edges);
        std::cout << "  All-pairs shortest paths:\n";
        for (int i = 0; i < n; ++i) {
            std::cout << "    from " << i << ": [";
            for (int j = 0; j < n; ++j) {
                if (j) std::cout << ", ";
                std::cout << (dist[i][j] >= FW_INF ? "INF" : std::to_string(dist[i][j]));
            }
            std::cout << "]\n";
        }
        assert(dist[0][2] == 5);  // 0->1(3)->2(2) = 5
        assert(dist[2][1] == 6);  // 2->3(1)->0(2)->1(3) = 6
        std::cout << "\n";
    }

    // ---- 4. DSU ----
    {
        std::cout << "-- DSU / Union-Find --\n";
        DSU dsu(6);
        dsu.unite(0, 1);
        dsu.unite(2, 3);
        dsu.unite(0, 2);
        assert(dsu.connected(1, 3));
        assert(!dsu.connected(0, 5));
        dsu.unite(4, 5);
        dsu.unite(3, 5);
        assert(dsu.connected(0, 5));
        std::cout << "  connected(1,3)=true, connected(0,5) after unions=true\n\n";
    }

    // ---- 5 & 6. Kruskal & Prim MST ----
    {
        std::cout << "-- Kruskal's MST --\n";
        //     0
        //    / \
        //   1   4
        //  / \ / \
        // 1   2   3
        //  \ | / |
        //   3   5
        int n = 6;
        std::vector<Edge> edges = {
            {0, 1, 4}, {0, 2, 4}, {1, 2, 2},
            {1, 3, 6}, {2, 3, 8}, {2, 4, 1},
            {3, 4, 7}, {3, 5, 9}, {4, 5, 3}
        };

        auto [kw, ke] = kruskal_mst(n, edges);
        std::cout << "  Kruskal total weight: " << kw << "\n";
        std::cout << "  edges: ";
        for (auto& [u, v, w] : ke) std::cout << "(" << u << "-" << v << ":" << w << ") ";
        std::cout << "\n";
        // Edges sorted by weight: (2,4,1), (1,2,2), (4,5,3), (0,1,4), (0,2,4), (1,3,6), ...
        // Pick: 2-4(1), 1-2(2), 4-5(3), 0-1(4), 1-3(6) = 1+2+3+4+6 = 16
        assert(kw == 16);

        std::cout << "\n-- Prim's MST --\n";
        AdjList adj(n);
        for (auto& [u, v, w] : edges) {
            adj[u].emplace_back(v, w);
            adj[v].emplace_back(u, w);
        }
        auto [pw, pe] = prim_mst(adj);
        std::cout << "  Prim total weight: " << pw << "\n";
        assert(pw == kw);  // Both should give same total weight.
        std::cout << "  (Kruskal == Prim: " << std::boolalpha << (pw == kw) << ")\n";
    }

    std::cout << "\nAll Week 22 tests passed.\n";
    return 0;
}
